plugins/chat/lib/chat/channel_archive_service.rb

```diff
@@ -107,22 +108,130 @@ def execute
         # Another future improvement is to send a MessageBus message for each
         # completed batch, so the UI can receive updates and show a progress
         # bar or something similar.
+
+        def create_post_from_batch(chat_messages, batch_thread_ranges)
+          create_post(
+            Chat::TranscriptService.new(
+              chat_channel,
+              chat_channel_archive.archived_by,
+              messages_or_ids: chat_messages,
+              thread_ranges: batch_thread_ranges,
+              opts: {
+                no_link: true,
+                include_reactions: true,
+              },
+            ).generate_markdown,
+          ) { delete_message_batch(chat_messages.map(&:id)) }
+        end
+
+        buffer = []
+        thread_ranges = {}
+        batch_thread_ranges = {}
+
         chat_channel
           .chat_messages
-          .find_in_batches(batch_size: ARCHIVED_MESSAGES_PER_POST) do |chat_messages|
-            create_post(
-              Chat::TranscriptService.new(
-                chat_channel,
-                chat_channel_archive.archived_by,
-                messages_or_ids: chat_messages,
-                opts: {
-                  no_link: true,
-                  include_reactions: true,
-                },
-              ).generate_markdown,
-            ) { delete_message_batch(chat_messages.map(&:id)) }
+          .order("created_at ASC")
+          .find_in_batches(batch_size: ARCHIVED_MESSAGES_PER_POST) do |message_batch|
+            thread_ids = message_batch.map(&:thread_id).compact.uniq
+            threads =
+              chat_channel
+                .chat_messages
+                .where(
+                  thread_id:
+                    Chat::Message
+                      .select(:thread_id)
+                      .where(thread_id: thread_ids)
+                      .group(:thread_id)
+                      .having("count(*) > 1"),
+                )
+                .order("created_at ASC")
+                .to_a
+
+            full_batch = (buffer + message_batch + threads).uniq { |msg| msg.id }
+            message_chunk = full_batch.group_by { |msg| msg.thread_id || msg.id }.values.flatten
+
+            buffer.clear
```

Comment:
what are we clearing here, is there any way there can be something in buffer at this point?

Reply:
It'is necessary to avoid carrying over the messages that didn't fit in the last post. So, we need to remove them from that buffer after adding them to the next batch

Reply:
OK then we will add comments